src/coreclr/jit/alloc.h

```diff
@@ -235,6 +235,16 @@ class CompAllocator
 
         void* p = m_arena->allocateMemory(count * sizeof(T));
 
+#if defined(_MSC_VER)
+        // MSVC does not handle the other pattern correctly to recognize this fact.
+        __assume(p != nullptr);
+#elif !defined(DEBUG)
+        if (p == nullptr)
+        {
+            UNREACHABLE();
+        }
+#endif
```

Comment:
Should there be a general `ASSUME(expr)` helper that abstracts this?

Reply:
It turns out there is already a `COMPILER_ASSUME` next to `UNREACHABLE` that ends up being `__assume` with the specified expression on MSVC, so I've switched to that.